evergreen_lint/rules/tasks_for_variants.py

```diff
@@ -0,0 +1,107 @@
+from typing import List
+
+from evergreen_lint.model import LintError, Rule
+
+
+class TasksForVariantsConfig(object):
+    def __init__(self, raw_mappings, failed_checks):
+        self._raw_mappings = raw_mappings
+        self._variant_mappings = {}
+        self._unused_variants = set()
+
+        for mapping in self._raw_mappings:
+            tasks = set(mapping["tasks"])
+            for variant in mapping["variants"]:
+                # Each variant can only be defined with one set of tasks.
+                if variant in self._variant_mappings:
+                    failed_checks.append(
+                        f"Invalid linter config: '{variant}' appeared more than once"
+                    )
+                self._variant_mappings[variant] = tasks
+                self._unused_variants.add(variant)
+
+    def tasks_for_variant(self, variant):
+
+        res = self._variant_mappings.get(variant, None)
+        if res is not None:
+            self._unused_variants.remove(variant)
+        return res
+
+    def assert_no_unused_variants(self, failed_checks):
+        if self._unused_variants:
+            failed_checks.append(
+                f"Invalid linter config: unknown variant names: {self._unused_variants}"
+            )
+
+
+class TasksForVariants(Rule):
+
+    """
+    Enforce task definitions for variants.
+
+    The configuration will look like:
+
+    ```
+    - rule: "tasks-for-variants"
+      task-variant-mappings:
+        - name: release-variants-tasks
+          tasks:
+          - task1
+          - task2
+          variants:
+          - variant1
+          - variant2
+        - name: asan-variant-tasks
+          tasks:
+          - task1
+          - task2
+          - task3
+          variants:
+          - variant3
+          - variant4
+    ```
+
+    """
+
+    @staticmethod
+    def name() -> str:
+        return "tasks-for-variants"
+
+    @staticmethod
+    def defaults() -> dict:
+        return {"task-variant-mappings": {}}
+
+    @staticmethod
+    def _get_task_set_from_list(task_list):
+        return {task["name"] for task in task_list}
+
+    def __call__(self, config: dict, yaml: dict) -> List[LintError]:
+        error_msg = (
+            "Mismatched task list for variant '{variant}'. Expected '{expected}', got '{actual}'"
+        )
+
+        failed_checks = []
+
+        mappings = config.get("task-variant-mappings")
```

Comment:
Please validate the format of mappings here. Then use `typing.cast` to make sure mypy has the type info.

Reply:
I inlined this variable in the next line; the intention is for the `TasksForVariantsConfig` to be the object mapper of the YAML config; the raw YAML should not be used directly except this one place.

The config object has type info per your comments above. There's minimal input validation; I expect contributors to deduce the format using existing rules.